pkg/traceql/ast_execute.go

```diff
@@ -100,76 +103,54 @@ func (o SpansetOperation) evaluate(input []*Spanset) (output []*Spanset, err err
 				output = append(output, matchingSpanset)
 			}
 
+		// relationship operators all set relFn which is used by below code
+		// to perform the operation
 		case OpSpansetDescendant:
-			spans, err := o.joinSpansets(lhs, rhs, func(l, r Span) bool {
+			fallthrough
+		case OpSpansetNotDescendant:
+			relFn = func(l, r Span) bool {
 				return r.DescendantOf(l)
-			})
-			if err != nil {
-				return nil, err
-			}
-
-			if len(spans) > 0 {
-				// Clone here to capture previously computed aggregates, grouped attrs, etc.
-				// Copy spans to new slice because of internal buffering.
-				matchingSpanset := input[i].clone()
-				matchingSpanset.Spans = append([]Span(nil), spans...)
-				output = append(output, matchingSpanset)
 			}
+			falseForAll = o.Op == OpSpansetNotDescendant
 
 		case OpSpansetAncestor:
-			spans, err := o.joinSpansets(lhs, rhs, func(l, r Span) bool {
-				// In case of ancestor the lhs becomes descendant of rhs
+			fallthrough
+		case OpSpansetNotAncestor:
+			relFn = func(l, r Span) bool {
 				return l.DescendantOf(r)
-			})
-			if err != nil {
-				return nil, err
-			}
-
-			if len(spans) > 0 {
-				// Clone here to capture previously computed aggregates, grouped attrs, etc.
-				// Copy spans to new slice because of internal buffering.
-				matchingSpanset := input[i].clone()
-				matchingSpanset.Spans = append([]Span(nil), spans...)
-				output = append(output, matchingSpanset)
 			}
+			falseForAll = o.Op == OpSpansetNotAncestor
 
 		case OpSpansetChild:
-			spans, err := o.joinSpansets(lhs, rhs, func(l, r Span) bool {
+			fallthrough
+		case OpSpansetNotChild:
+			relFn = func(l, r Span) bool {
 				return r.ChildOf(l)
-			})
-			if err != nil {
-				return nil, err
-			}
-
-			if len(spans) > 0 {
-				// Clone here to capture previously computed aggregates, grouped attrs, etc.
-				// Copy spans to new slice because of internal buffering.
-				matchingSpanset := input[i].clone()
-				matchingSpanset.Spans = append([]Span(nil), spans...)
-				output = append(output, matchingSpanset)
 			}
+			falseForAll = o.Op == OpSpansetNotChild
 
 		case OpSpansetParent:
-			spans, err := o.joinSpansets(lhs, rhs, func(l, r Span) bool {
-				// In case of parent the lhs becomes child of rhs
+			fallthrough
+		case OpSpansetNotParent:
+			relFn = func(l, r Span) bool {
 				return l.ChildOf(r)
-			})
-			if err != nil {
-				return nil, err
-			}
-
-			if len(spans) > 0 {
-				// Clone here to capture previously computed aggregates, grouped attrs, etc.
-				// Copy spans to new slice because of internal buffering.
-				matchingSpanset := input[i].clone()
-				matchingSpanset.Spans = append([]Span(nil), spans...)
-				output = append(output, matchingSpanset)
 			}
+			falseForAll = o.Op == OpSpansetNotParent
```

Comment:
Could the falseForAlls be set in the prior cases before the fallthrough?

Reply:
if we set it before the fallthrough won't it just get overwritten in the next case? i might be confused about what you're asking for. Something like?
```
		case OpSpansetDescendant:
			falseForAll = false
			fallthrough
		case OpSpansetNotDescendant:
                        falseForAll = true
			relFn = <some func>
```
we could init it to `true` and then set to false in all of the cases like `OpSpansetDescendant` ?

Reply:
Was thinking something like:
```go
		case OpSpansetNotDescendant:
			falseForAll = true
			fallthrough
		case OpSpansetDescendant:
			relFn = <some func>
```

I didn't check the order of each case pair

Reply:
oh. yeah. that does read better.